packages/shacl-language-server/src/ShaclLanguageServer.ts

```diff
@@ -0,0 +1,226 @@
+import * as lsp from 'vscode-languageserver';
+import uniqBy from 'lodash.uniqby';
+import { autoBindMethods } from 'class-autobind-decorator';
+import {
+  errorMessageProvider,
+  AbstractLanguageServer,
+  CompletionCandidate,
+  regexPatternToString,
+  getTokenTypesForCategory,
+} from 'stardog-language-utils';
+import { ShaclParser, shaclTokens } from 'millan';
+import { Lexer, TokenType } from 'chevrotain';
+
+const SHACL_TOKEN_PREFIX = 'SHACL_';
+const PREFIXED_SUFFIX = '_prefixed';
+const IRI_SUFFIX = '_IRI';
+const shaclTokenMap = shaclTokens.getShaclTokenMap({
+  // TODO in future: put this inside of ShaclLanguageServer and allow client to
+  // specify arbitrary namespaces. This is already supported by `millan`'s
+  // parser.
+  shacl: 'sh',
+  xsd: 'xsd',
+});
+const baseShaclTokens = Object.keys(shaclTokenMap).reduce(
+  (accumulator, key) =>
+    key.endsWith(PREFIXED_SUFFIX) || key.endsWith(IRI_SUFFIX)
+      ? accumulator
+      : accumulator.concat(shaclTokenMap[key]),
+  [] as TokenType[]
+);
+
+@autoBindMethods
+export class ShaclLanguageServer extends AbstractLanguageServer<ShaclParser> {
+  constructor(connection: lsp.IConnection) {
+    super(
+      connection,
+      new ShaclParser(
+        { errorMessageProvider },
+        {
+          shacl: 'sh',
+          xsd: 'xsd',
+        }
+      )
+    );
+  }
+
+  onInitialization(_params: lsp.InitializeParams): lsp.InitializeResult {
+    this.connection.onCompletion(this.handleCompletion);
+
+    return {
+      capabilities: {
+        // Tell the client that the server works in NONE text document sync mode
+        textDocumentSync: this.documents.syncKind[0],
+        completionProvider: {
+          triggerCharacters: ['<', ':'],
+        },
+        hoverProvider: true,
+      },
+    };
+  }
+
+  onContentChange(
+    { document }: lsp.TextDocumentChangeEvent,
+    parseResults: ReturnType<
+      AbstractLanguageServer<ShaclParser>['parseDocument']
+    >
+  ) {
+    const { uri } = document;
+    const content = document.getText();
+    const { errors, tokens, otherParseData } = parseResults;
+
+    if (!content.length) {
+      this.connection.sendDiagnostics({
+        uri,
+        diagnostics: [],
+      });
+      return;
+    }
+
+    const lexDiagnostics = this.getLexDiagnostics(document, tokens);
+    const parseDiagnostics =
+      otherParseData.semanticErrors.length > 0
```

Comment:
Should we check for the existence of `semanticErrors` first?

Reply:
No, it's guaranteed to exist on `parseResults` (not optional). The types will show this:

![image](https://user-images.githubusercontent.com/8409748/59697947-b51d8280-91bc-11e9-8f07-4a237d14b976.png)